Framework/Geometry/src/Objects/MeshObject.cpp

```diff
@@ -0,0 +1,690 @@
+  }
+  m_material = Mantid::Kernel::make_unique<Material>(material);
+  m_handler = boost::make_shared<CacheGeometryHandler>(this);
+}
+
+/**
+ * @return The Material that the object is composed from
+ */
+const Kernel::Material MeshObject::material() const {
+  if (m_material)
+    return *m_material;
+  else
+    return Material();
+}
+
+/**
+* Returns whether this object has a valid shape
+* @returns True if the entire MeshObject may enclose
+* one or more volumes.
+*/
+bool MeshObject::hasValidShape() const {
+  // May enclose volume if there are at
+  // at least 4 triangles and 4 vertices
+  return (numberOfTriangles() >= 4 && numberOfVertices() >= 4);
+}
+
+/**
+* Determines whether point is within the object or on the surface
+* @param point :: Point to be tested
+* @returns true if point is within object or on surface
+*/
+bool MeshObject::isValid(const Kernel::V3D &point) const {
+
+  BoundingBox bb = getBoundingBox();
+  if (!bb.isPointInside(point)) {
+    return false;
+  }
+
+  V3D direction(0.0, 0.0, 1.0); // direction to look for intersections
+  std::vector<V3D> intesectionPoints;
+  std::vector<int> entryExitFlags;
+
+  getIntersections(point, direction, intesectionPoints, entryExitFlags);
+
+  if (intesectionPoints.size() == 0) {
+    return false;
+  }
+
+  // True if point is on surface
+  for (size_t i = 0; i < intesectionPoints.size(); ++i) {
+    if (point.distance(intesectionPoints[i]) < M_TOLERANCE) {
+      return true;
+    }
+  }
+
+  // Look for nearest point then check its entry-exit flag
+  double nearestPointDistance = point.distance(intesectionPoints[0]);
+  size_t nearestPointIndex = 0;
+  for (size_t i = 1; i < intesectionPoints.size(); ++i) {
+    if (point.distance(intesectionPoints[i]) < nearestPointDistance) {
+      nearestPointDistance = point.distance(intesectionPoints[i]);
+      nearestPointIndex = i;
+    }
+  }
+  return (entryExitFlags[nearestPointIndex] == -1);
+}
+
+/**
+* Determines wither point is on the surface.
+* @param point :: Point to check
+* @returns true if the point is on the surface
+*/
+bool MeshObject::isOnSide(const Kernel::V3D &point) const {
+
+  BoundingBox bb = getBoundingBox();
+  if (!bb.isPointInside(point)) {
+    return false;
+  }
+
+  std::vector<V3D> directions; // directions to look for intersections
+  directions.push_back(V3D(0, 0, 1));
+  directions.push_back(V3D(0, 1, 0));
+  directions.push_back(V3D(1, 0, 0));
+  // We have to look in several directions in case a point is on a face
+  // or edge parallel to the first direction or also the second direction.
+  for (size_t i = 0; i < directions.size(); ++i) {
+    std::vector<V3D> intesectionPoints;
+    std::vector<int> entryExitFlags;
+
+    getIntersections(point, directions[i], intesectionPoints, entryExitFlags);
+
+    if (intesectionPoints.size() == 0) {
+      return false;
+    }
+
+    size_t k = intesectionPoints.size();
+    for (size_t i = 0; i < k; ++i) {
```

Comment:
You are already using `i` in the outer loop, but this would be solved by making the outer loop range based.